pkg/distributor/forwarding/forwarding.go

```diff
@@ -96,185 +76,286 @@ func NewForwarder(reg prometheus.Registerer, cfg Config) Forwarder {
+	if err != nil {
+		// Can't check whether a timeseries should be forwarded if it has no metric name.
+		// Ingest it and don't forward it.
+		return "", true
 	}
 
-	go func() {
-		defer r.cleanup()
-		defer close(errCh)
+	rule, ok := rules[metric]
+	if !ok {
+		// There is no forwarding rule for this metric, ingest it and don't forward it.
+		return "", true
+	}
 
-		wg.Wait()
+	return rule.Endpoint, rule.Ingest
+}
 
-		var nonRecoverable error
-		// No need to get errorsMtx because we already waited for all routines which might modify it to end.
-		for endpoint, err := range errorsByEndpoint {
-			if err == nil {
-				continue
-			}
+type request struct {
+	pools  *pools
+	client *http.Client
+	log    log.Logger
 
-			if errors.As(err, &recoverableError{}) {
-				// If there is at least one recoverable error we want to return the recoverable error.
-				returnErr(httpgrpc.Errorf(http.StatusInternalServerError, "endpoint %s: %s", endpoint, err.Error()))
-				return
-			}
+	ctx             context.Context
+	timeout         time.Duration
+	propagateErrors bool
+	errCh           chan error
+	requestWg       *sync.WaitGroup
 
-			nonRecoverable = httpgrpc.Errorf(http.StatusBadRequest, "endpoint %s: %s", endpoint, err.Error())
-		}
+	endpoint string
+	ts       tsWithSampleCount
 
-		if nonRecoverable != nil {
-			returnErr(nonRecoverable)
-		}
-	}()
+	requests  prometheus.Counter
+	errors    *prometheus.CounterVec
+	samples   prometheus.Counter
+	exemplars prometheus.Counter
+	latency   prometheus.Histogram
+}
 
-	return errCh
+// submitForwardingRequest launches a new forwarding request and sends it to a worker via a channel.
+// It might block if all the workers are busy.
+func (f *forwarder) submitForwardingRequest(ctx context.Context, endpoint string, ts tsWithSampleCount, requestWg *sync.WaitGroup, errCh chan error) {
+	req := f.pools.getReq()
+
+	req.pools = f.pools
+	req.client = &f.client // http client should be re-used so open connections get re-used.
+	req.log = f.log
+	req.ctx = ctx
+	req.timeout = f.cfg.RequestTimeout
+	req.propagateErrors = f.cfg.PropagateErrors
+	req.errCh = errCh
+	req.requestWg = requestWg
+
+	// Target endpoint and TimeSeries to forward.
+	req.endpoint = endpoint
+	req.ts = ts
+
+	// Metrics.
+	req.requests = f.requestsTotal
+	req.errors = f.errorsTotal
+	req.samples = f.samplesTotal
+	req.exemplars = f.exemplarsTotal
+	req.latency = f.requestLatencyHistogram
+
+	select {
+	case <-ctx.Done():
+	case f.reqCh <- req:
+	}
 }
 
-// sendToEndpoint sends the given timeseries to the given endpoint.
-// All returned errors which are recoverable are of the type recoverableError.
-func (r *request) sendToEndpoint(ctx context.Context, endpoint string, ts []mimirpb.PreallocTimeseries) error {
-	protoBufBytes := (*r.pools.protobuf.Get().(*[]byte))[:0]
+// do performs a forwarding request.
+func (r *request) do() {
+	defer r.cleanup()
+
+	protoBufBytes := (*r.pools.getProtobuf())[:0]
+	defer r.pools.putProtobuf(&protoBufBytes)
```

Comment:
Non-blocking: I wonder if by saving the original reference we can avoid creating a new one.

Right now `r.pools.getProtobuf()` gives you a location in memory from where you're copying the slice header into `protoBufBytes` and throwing away the reference. However `protoBufBytes` also escapes to heap since you're returning a reference in the defer.

What if we do something like:
```suggestion
	protoBufBytesRef := r.pools.getProtobuf()
	protoBufBytes := (*protoBufBytesRef)[:0]
	defer func() {
		*protoBufBytesRef = protoBufBytes // just in case we increased its capacity
		r.pools.putProtobuf(protoBufBytesRef)
	}()
```

Reply:
interesting idea!

I benchmarked it, the results were a bit inconclusive because I don't understand why only one of the two benchmark cases that are performing forwarding requests see an improvement, but if it "might" improve the allocations then I think we can apply this suggestion because the down-side that comes from the additional complexity is minimal (I don't think it's very complicated).

```
replay@mauro-workstation:~/go/src/github.com/grafana/mimir$ benchstat -delta-test none ./before ./after
name                                                      old time/op    new time/op    delta
RemoteWriteForwarding/forwarding_and_ingesting-4            1.66ms ± 0%    1.70ms ± 0%   +2.63%
RemoteWriteForwarding/forwarding_and_no_ingesting-4         1.76ms ± 0%    1.90ms ± 0%   +7.51%
RemoteWriteForwarding/no_forwarding_and_only_ingesting-4    1.44ms ± 0%    1.47ms ± 0%   +2.39%

name                                                      old alloc/op   new alloc/op   delta
RemoteWriteForwarding/forwarding_and_ingesting-4            2.43kB ± 0%    2.49kB ± 0%   +2.43%
RemoteWriteForwarding/forwarding_and_no_ingesting-4         2.53kB ± 0%    2.79kB ± 0%  +10.35%
RemoteWriteForwarding/no_forwarding_and_only_ingesting-4      190B ± 0%      192B ± 0%   +1.05%

name                                                      old allocs/op  new allocs/op  delta
RemoteWriteForwarding/forwarding_and_ingesting-4              29.0 ± 0%      28.0 ± 0%   -3.45%
RemoteWriteForwarding/forwarding_and_no_ingesting-4           29.0 ± 0%      29.0 ± 0%    0.00%
RemoteWriteForwarding/no_forwarding_and_only_ingesting-4      5.00 ± 0%      5.00 ± 0%    0.00%
```